### PR TITLE
Added per-profile tuning step, made receiver follow tuning step, made…

### PR DESCRIPTION
**Pull request by luarvique**
Changes:

1) Added per-profile tuning step setting.
2) Made the main receiver interface follow the tuning step, rounding any tuned frequency to the nearest step.
3) Made mouse scroll wheel tune the frequency (in tuning steps) rather than zoom.
4) Retained zoom functionality, now activated if the user is pushing scroll wheel down, or holding the right mouse button down.
5) Added filter boundaries display.

### CODE REVIEW

**Review by jketterl — CHANGES_REQUESTED** [review]
There's two major things that prevent this from being merged:

* The frequency step selection can only be done by the operator and cannot be configured or disabled by the user. This is not acceptable.
* The change of the zoom operation to require an additional mouse button is a significant change that will require proper qualification. I think that the mousewheel zoom is an essential part of the user experience in pretty much any SDR software, so I'd say that if there needs to be an additional tuning function on the mouse wheel, it should be the one that requires a modifier button. Also note that mousewheel tuning is already available if the mouse cursor hovers over the frequency display. There should probably also be a discussion on whether mouse buttons are reasonable to use as modifiers.

Since you don't seem to have a lot of information about this project, nor seem to be very eager to read up on its past: There has been previous attempts to implement frequency stepping in the receiver. The major restriction here is that I don't want the receiver to be spammed with additional buttons for minor features (I do consider this a minor feature).

File: owrx/source/__init__.py
Comment (jketterl) [inline owrx/source/__init__.py:573]: The `Option` class probably does not lend itself very well to the use with numeric values (which probably also explains the need to use an `IntConverter`). This would probably work better with an enum.

Comment (luarvique) [reply owrx/source/__init__.py:573]: This pattern has been copied from the existing code for the "tuning precision" setting. It can probably be refactored to something better, but I tend to stick to the original coding style when making small changes.

File: owrx/source/__init__.py
Comment (luarvique) [reply owrx/source/__init__.py:598]: There is no forcing going on here. It defaults to 1Hz, i.e. the original behavior. You can ask around though and easily find out that nobody needs to tune in 1Hz increments, even CW aficionados.

File: owrx/config/defaults.py
Comment (luarvique) [reply owrx/config/defaults.py:148]: The defaults are set to reasonable values anyone would expect for these bands. You can see the same steps in HDSDR and other packages.

File: owrx/config/defaults.py
Comment (luarvique) [reply owrx/config/defaults.py:36]: Will fix.

File: owrx/source/__init__.py
Comment (jketterl) [reply owrx/source/__init__.py:598]: You're forcing every receiver operator to fill this field. If the default is 1Hz, it can as well be absent and be added when necessary, thus avoiding unnecessary confusion.

File: htdocs/openwebrx.js
Comment (luarvique) [reply htdocs/openwebrx.js:812]: Will fix.

File: htdocs/openwebrx.js
Comment (luarvique) [reply htdocs/openwebrx.js:34]: This is done in sync with the original code style.

File: htdocs/openwebrx.js
Comment (luarvique) [reply htdocs/openwebrx.js:562]: Very well, I will invert the behavior, so that tuning is only available with the scroll wheel pushed (or right button pressed). Will this satisfy your desire not to change the current behavior of the interface?

File: htdocs/openwebrx.js
Comment (jketterl) [reply htdocs/openwebrx.js:34]: Yeah, but the original code is beyond reasonable in that aspect. I have been working at restructuring it for years now. I don't appreciate you adding to the pile of cleanup that I still have to work through.

**Comment by luarvique** [comment]
_The frequency step selection can only be done by the operator and cannot be configured or disabled by the user. This is not acceptable._
This does not prevent the operator to keep step at the default 1Hz value, while giving him an option to set it to a different value. So, this is definitely better than not having any options at all. Why is this not acceptable? 

_The change of the zoom operation to require an additional mouse button is a significant change that will require proper qualification. I think that the mousewheel zoom is an essential part of the user experience in pretty much any SDR software._
Very well. I will invert the behavior, so that the essential part of the user experience still happens by default then, with no extra buttons pressed.

_Since you don't seem to have a lot of information about this project, nor seem to be very eager to read up on its past: There has been previous attempts to implement frequency stepping in the receiver. The major restriction here is that I don't want the receiver to be spammed with additional buttons for minor features (I do consider this a minor feature)._

As to receiver being spammed with additional buttons, there are no buttons being added by these changes. There is a single additional tuning step setting associated with each profile. The tuning step can be found in pretty much every other SDR package out there, so it is somewhat surprising that it is not present in OpenWebRX. Must be because of the major restriction.

While I do not have a lot of information about this project, I do have enough information and prior experience to understand what is going on here. This is not new, happened many times before in many other projects.

File: htdocs/openwebrx.js
Comment (luarvique) [reply htdocs/openwebrx.js:34]: If I start making changes in a different style, you will get a very messy pile to cleanup. But, if you insist, please provide me with an example of how you want this variable to look in the code and I will put there whatever you want it to be.

File: owrx/source/__init__.py
Comment (luarvique) [reply owrx/source/__init__.py:598]: Since the field comes up with a default value, I am not forcing operator to do anything here. I am just making that field *visible* to the operator, which is a logical thing to do given that very, very few people expect to tune in 1Hz steps. Making visible is different from forcing, would you agree?

Comment (jketterl) [reply owrx/source/__init__.py:598]: Well then please explain to me why making the field visible is required. I still don't see why this should be a required field on all devices and profiles.

Also, making the field mandatory prevents the operator from using the inheritance (which would allow a default setting on the device, with the profiles automatically inheriting that setting).

File: owrx/source/__init__.py
Comment (jketterl) [reply owrx/source/__init__.py:573]: The tuning precision is implemented a bit different from this since it uses a `range()` instead of an array of predetermined values. The latter is closer to an enum for me.

The important part here is that the config contains numbers, and the default configuration containing strings concerned me a little in this place. If that part works, it's not super-important how this is modeled since this is just about the visual representation while editing the settings.

Comment (jketterl) [reply owrx/source/__init__.py:573]: Also, one more thing I just noticed: The list should probably also contain some more options:

* 9kHz (MW broadcast band)
* 12.5kHz and 25kHz (ham radio narrow band channel spacing)
* 100kHz (broadcast FM band)

File: htdocs/openwebrx.js
Comment (jketterl) [reply htdocs/openwebrx.js:34]: That's why I've been saying that the frontend code needs a proper refactoring before adding new features. You can have a look at `htdocs/lib/DemodulatorPanel.js` for an example of how that would look like. The problem is that you can't  just remove the need for global variables from this code without restructuring it completely.

Keep the variable, but please note the need for proper structure here.

File: owrx/source/__init__.py
Comment (luarvique) [reply owrx/source/__init__.py:598]: This field needs to be visible because in most cases the operator will want to set it. I do not know if you ever actually listen to a radio, but everyone who has knows that unless you tune to the exact frequency of the transmitter, your audio becomes garbled. And most transmitters (commercial or amateur) operate at certain steps, depending on the band and modulation. Tuning outside these steps, while occasionally useful, does not provide usable listening experience.

As to preventing inheritance, profiles already have the modulation setting that is also not present in the device settings. This does not prevent inheritance. The tuning step should be treated in pretty much the same way as modulation.

To make this argument more visual, I am going to include screenshots of several SDR packages for your viewing pleasure, showing tuning step right in front of the UI.

HDSDR:
image

SDR Console:
image

SDR Sharp:
image

**Comment by luarvique** [comment]
_Because the user should have the ability to override this setting. It certainly isn't a problem if the setting is at 1Hz, but the operator may always set a higher setting for various erroneous reasons (lack of foresight, neglect, lack of knowledge) leaving the user unable to tune signals that don't fit the grid._

While it would be *nice* to let the user override this setting, having at least operator-modifiable setting is *better than no setting at all*. As to the argument that the operator "may make a mistake", there are plenty of places where operator may make a mistake, this should not be the reason to take functionality from the operator.

_This was a heads-up simply because that's what has happened before. I know that there's no buttons added right now, but since I am also requiring a way for the user to control this feature, it will probably end in this direction. Would you prefer me to wait until you have added the buttons to tell you this?_

I would prefer if you stopped inventing reasons not to have tuning step functionality and just included it into the software, either implemented by yourself or implemented by me. If you decide to add user UI for this functionality later, all the best to you. If you decide not to add user UI, at least the operator would be able to configure it. If the operator does not bother to configure it, then in my implementation it safely stays at 1Hz, keeping the original behavior.

_How do you think this project ended up in my hands?_

Let me guess... The previous maintainer had certain highly specific mental blocks and acted on them, rejecting any external contributions? At some point, contributors stopped coming and the project was eventually abandoned?

File: owrx/source/__init__.py
Comment (jketterl) [reply owrx/source/__init__.py:598]: > This field needs to be visible because in most cases the operator will want to set it. I do not know if you ever actually listen to a radio, but everyone who has knows that unless you tune to the exact frequency of the transmitter, your audio becomes garbled. And most transmitters (commercial or amateur) operate at certain steps, depending on the band and modulation. Tuning outside these steps, while occasionally useful, does not provide usable listening experience.

This may be true for listening to broadcast stations, but this is not the only use case. When listening to ham broadcasts in SSB or CW you're better off with a more fine-grained control, and I don't think these modes profit from the tuning step at all. As such I don't agree with the opinion that operators will want to set this on every profile.

Also, please note that both transmitters as well as receivers may be off frequency (depending on their quality of calibration), so tuning to a presumably exact frequency may actually not yield the optimal result.

> As to preventing inheritance, profiles already have the modulation setting that is also not present in the device settings. This does not prevent inheritance. The tuning step should be treated in pretty much the same way as modulation.

I see, the setting is only available for profiles, so there's no inheritance, making this point moot. That then of course raises the question of whether operators could profit from enabling the inheritance.

> To make this argument more visual, I am going to include screenshots of several SDR packages for your viewing pleasure, showing tuning step right in front of the UI.

I'm sorry, but the fact that others do it some way has never been a good argument.

File: htdocs/openwebrx.js
Comment (luarvique) [reply htdocs/openwebrx.js:562]: Inverted in the last commit.

File: htdocs/openwebrx.js
Comment (luarvique) [reply htdocs/openwebrx.js:814]: Fixed in the last commit.

File: owrx/config/defaults.py
Comment (luarvique) [reply owrx/config/defaults.py]: Fixed in the last commit.

File: owrx/source/__init__.py
Comment (luarvique) [reply owrx/source/__init__.py:598]: _This may be true for listening to broadcast stations, but this is not the only use case. When listening to ham broadcasts in SSB or CW you're better off with a more fine-grained control, and I don't think these modes profit from the tuning step at all. As such I don't agree with the opinion that operators will want to set this on every profile._
SSB tuning is usually 500Hz, CW is 100Hz. Both tuning steps are present in my changes.

_I'm sorry, but the fact that others do it some way has never been a good argument._
Mmmm... ok. Then consider it a reason to _think_ what exactly makes all these others do it this way.

**Comment by jketterl** [comment]
> While it would be _nice_ to let the user override this setting, having at least operator-modifiable setting is _better than no setting at all_. As to the argument that the operator "may make a mistake", there are plenty of places where operator may make a mistake, this should not be the reason to take functionality from the operator.

I disagree. I'd say the user interface is actually more important than giving the operator the ability to select the tuning step on behalf of the user.

> I would prefer if you stopped inventing reasons not to have tuning step functionality and just included it into the software, either implemented by yourself or implemented by me. If you decide to add user UI for this functionality later, all the best to you. If you decide not to add user UI, at least the operator would be able to configure it. If the operator does not bother to configure it, then in my implementation it safely stays at 1Hz, keeping the original behavior.

Who knows, if I wasn't confronted with so many impatient people that want to shove their opinions and solutions down my throat, we may actually have a solution already. Collaboration should be the goal here, but whenever I bring the actual problems to the table, things quickly go downhill. People start ignoring me and keep on pushing their own agenda, painting me as the problem. I don't see this as a viable project environment, but when has that ever mattered? The mental state of a project maintainer must be able to take unlimited strain, and he must always immediately take the time to answer a user's request...

I don't know if you know, but the main reason why I don't want this application to look like others is because I'd like to retain the ease of use for radio beginners. I'd say that a websdr has the best accessibility for any user these days (given that a browser with internet access has become very common these days) and as such I consider absolute beginners at least as part of the target audience. This is the main reason why I am looking to build a very intuitive user interface (which is something I miss in many other SDR packages). I am not very good at it, and many others that come into contact with the project for the first time don't take the time to understand or work on that challenge, they just want their feature in the software, right now, and I'm sick of it.

There, I hope that was enough of an emotional disclosure.

> Let me guess... The previous maintainer had certain highly specific mental blocks and acted on them, rejecting any external contributions? At some point, contributors stopped coming and the project was eventually abandoned?

Not quite, I would say that Andras (the previous maintainer) simply moved on and did not have the time to maintain the project any more.

File: owrx/source/__init__.py
Comment (luarvique) [reply owrx/source/__init__.py:573]: Yeah, I will add these.

**Comment by luarvique** [comment]
_I disagree. I'd say the user interface is actually more important than giving the operator the ability to select the tuning step on behalf of the user._

Sorry, but this makes no sense. You _can_ have tuning step functionality without end user interface. You _cannot_ have end user interface without the functionality. In other words, if you consider user interface important, you can always add it later, adding to the functionality I already implemented for you.

_Who knows, if I wasn't confronted with so many impatient people that want to shove their opinions and solutions down my throat, we may actually have a solution already. Collaboration should be the goal here, but whenever I bring the actual problems to the table, things quickly go downhill. People start ignoring me and keep on pushing their own agenda, painting me as the problem._

Well, I have been very patient so far, politely stating the issue, implementing the solution with the minimal amount of the code base changes, presenting my changes to the maintainer, even going as far as using visual aids.

You obviously do not _need_ to accept changes, but if you do not, all these pesky, impatient people will just stop coming and the project will eventually wither.

_I don't know if you know, but the main reason why I don't want this application to look like others is because I'd like to retain the ease of use for radio beginners. I'd say that a websdr has the best accessibility for any user these days (given that a browser with internet access has become very common these days) and as such I consider absolute beginners at least as part of the target audience._ 

I actually consider OpenWebRX *way better* than WebSDR in terms of both UI and functionality, but do not understand the need to dumb it down "for beginners".

It is true that the UI has to be simple and intuitive (see SDRuno for a negative example). But this can be done by hiding more complex functionality from casual users rather than taking it away. The user / operator division in OpenWebRX seems to be a natural point to do that, with more functionality revealed to users as you devise better ways to present it or create different experience levels for different users.

Then, there is the issue of basic SDR functionality, i.e. "what other people do". Skipping on the basic functionality, just not to be like "the other people", is *not* a good idea, however independent and artistic it may sound. There is nothing wrong in looking at HDSDR and making sure your app has the same basic set of features, because that feature set is the result of years of development and usage.

File: owrx/source/__init__.py
Comment (luarvique) [reply owrx/source/__init__.py:573]: Done.

File: htdocs/openwebrx.js
Comment (luarvique) [reply htdocs/openwebrx.js:814]: Going to roll back the 'else' change, since this is not how on_ws_recv() function works. The config set it gets may contain any number of items. Items not passed via config set should not be modified in on_ws_recv() (see handling for other items from that set).

**Comment by jketterl** [comment]
> >I disagree. I'd say the user interface is actually more important than giving the operator the ability to select the tuning step on behalf of the user.
> 
> Sorry, but this makes no sense. You _can_ have tuning step functionality without end user interface. You _cannot_ have end user interface without the functionality. In other words, if you consider user interface important, you can always add it later, adding to the functionality I already implemented for you.
> 

What I meant of course was not just a user interface, but a working user interface, meaning that of course the functionality should also be present. Implementing the functionality is not the big deal.

I will once again try to make my point: Adding the tuning step selection for the operator without having a corresponding interface for the user represents a degradation in user experience to me, so as long as there's no integration of this into the receiver interface, this will simply join the waiting line.

You posted a series of screenshots illustrating the tuning step selector being present in other SDR software. Please note that in all of these screenshots, the tuning step selector is available _to the user_. Of course, the fact that there is no separate operator role in any of these examples makes it a bit hard to immediately come to a conclusion, but just imagine that that selector was not available and you could only tune in the steps predetermined by the developer.

> Well, I have been very patient so far, politely stating the issue, implementing the solution with the minimal amount of the code base changes, presenting my changes to the maintainer, even going as far as using visual aids.

No, I'm sorry, but I have certainly not experienced your communication as being patient or polite.

> You obviously do not _need_ to accept changes, but if you do not, all these pesky, impatient people will just stop coming and the project will eventually wither.

Maybe. It's impossible to predict the future though.

> > I don't know if you know, but the main reason why I don't want this application to look like others is because I'd like to retain the ease of use for radio beginners. I'd say that a websdr has the best accessibility for any user these days (given that a browser with internet access has become very common these days) and as such I consider absolute beginners at least as part of the target audience.
> 
> I actually consider OpenWebRX _way better_ than WebSDR in terms of both UI and functionality, but do not understand the need to dumb it down "for beginners".

I just explained it in the paragraph above.

> It is true that the UI has to be simple and intuitive (see SDRuno for a negative example). But this can be done by hiding more complex functionality from casual users rather than taking it away. The user / operator division in OpenWebRX seems to be a natural point to do that, with more functionality revealed to users as you devise better ways to present it or create different experience levels for different users.

I have never said that advanced features have no place. But, in contrast to other software, I am not just going to add everything that you could possible add just for the sake of being able to. The idea here is that the basic functionality must be retained for those that do not want to be bothered with a metric ton of features. The problem is that this presents a new user experience challenge that remains to be resolved.

> Then, there is the issue of basic SDR functionality, i.e. "what other people do". Skipping on the basic functionality, just not to be like "the other people", is _not_ a good idea, however independent and artistic it may sound. There is nothing wrong in looking at HDSDR and making sure your app has the same basic set of features, because that feature set is the result of years of development and usage.

I don't think that replicating another piece of software is a good strategy in general. If you think that another piece of software already has everything you need, it would be smarter to just use that other piece of software right from the start.

The only aspect that differentiates OpenWebRX from other SDR software (at the base architectural level) is the fact that it is built as a web application that runs in a browser. There is a lot of considerations stemming from that, though. For example: The whole separation of roles between user and operator stems from this underlying architectural difference. As such, I think it is only reasonable to re-evaluate existing features.

**Comment by jketterl** [comment]
Can we please handle your changes one at a time? The addition of a callsign lookup is a completely new and unrelated feature.

File: htdocs/openwebrx.js
Comment (jketterl) [reply htdocs/openwebrx.js:812]: You're right, I'm sorry. Mea culpa. I've even done this wrong myself...

Config changes are incremental, so indeed, the absence of a key does not mean that it should reset to the default.

I'll try to find the discussion about mandatory vs. optional keys first because this kind of hinges on that.

File: owrx/source/__init__.py
Comment (jketterl) [reply owrx/source/__init__.py:598]: Coming back to this, I think this discussion went in a completely wrong way. The question here should still be: is this a field that is required to achieve a working profile, and I'd say the answer to that is clearly: No.

The whole idea of having optional fields came from the original text-based configuration where the full extent of available keys could easily overwhelm operators. For that reason, many of the keys were optional and could be added to the configuration as needed after consulting the documentation.

This has been translated into the web configuration interface: It only shows the bare minimum, with the ability to add more functionality as needed. All the optional keys will have some kind of reasonable default.

This even applies to very common keys, for example "gain".

Comment (luarvique) [reply owrx/source/__init__.py:598]: For gain, you can safely leave it out and let it get derived from the device profile. The tuning step has nothing to do with the device profile and thus isn't present there. So, its default value cannot be derived from the device profile. It is supposed to be set based on the band frequency and modulation, i.e. specific to each band profile.

Since you are so eager to reduce the number of fields operator "needs" to enter, why do you have the starting frequency in the profile? What if I do not care about the starting frequency? Why not fix it to the center frequency by default, the same way you would like to keep the tuning step to 1Hz y default?

Comment (luarvique) [reply owrx/source/__init__.py:598]: I think the whole reduction of the configurable items is a topic of a separate discussion, since this dogmatic approach has ruined many products (GNOME, Windows 10, etc).

If you believe that a setting will be configured often, do not hide it. If you do not want to force the user to enter a value, provide a reasonable default value. This is exactly what has been done with the tuning step. In fact, I would insist on 1kHz as a reasonable default value, but to appease your sensibilities, it is set to 1Hz at the moment.

Comment (jketterl) [reply owrx/source/__init__.py:598]: I think that's the whole problem in this discussion: You expect this to be a feature that operators will want to use on all profiles, while I believe that many operators won't be bothered to set this up (correctly).

Comment (luarvique) [reply owrx/source/__init__.py:598]: Since there is a default value that retains the original behavior when not set, I see no problem with operators not bothering to set it.

Where I do see a problem is that tuning in 1Hz steps (the only option at the moment) makes precise tuning to a station nearly impossible without zooming the waterfall. And even with zoomed waterfall, it is often hit and miss. Thus is my conclusion that _most_ operators _will_ want to set the tuning step value.

Comment (jketterl) [reply owrx/source/__init__.py:598]: Your conclusion is not immediately wrong, but you're placing it into the wrong role. It's not the operators that want this feature, it's the users. This is confirmed by multiple requests on the project.

You're the first to involve the operator in this scenario, and while I do agree that it's not completely unreasonable to give the operator the ability to preselect a tuning step, it's important to understand that an operator should not be the one to make the final decision on this. As such, I find it's perfectly reasonable for an operator to skip this configuration and leave it completely up to the user.

Also, as I already stated previously, precision tuning is a myth. You will always have inaccuracies. You are using an rtl-sdr device yourself, right? It has a maximum precision of 1ppm, so even if you have done the calibration with utmost precision, you will still be up to 0.5 ppm off. Are you going to notice that? I doubt it.

Comment (luarvique) [reply owrx/source/__init__.py:598]: Look, I _can_ add this as a user-configurable option to the receiver panel, with the initial value configured by the operator (as done in the discussed change). The problem is that once I do this, you will reject the change claiming that I am cluttering the user interface. That is why I am not pushing any receiver panel changes, only sticking to safer, less visible operator interface changes.

Thus, for me, it is catch-22 situation. For you, it is "let the annoying freak run in circles, he is not getting in anyway". Alas, this is how it usually is with this kind of projects.

**Comment by jketterl** [comment]
I am going to close this PR since you just keep adding stuff, completely ignoring my previous request to keep features separated. This is not a reasonable workflow and only ends in a big mess.

**Pull request closed without merging**
